File: src/main/java/leet_code/SymmetricTree.java
package leet_code;

import leet_code.SameTree.TreeNode;

public class SymmetricTree {

    public static void main(String[] args) {

    }


    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return mirror(root.left, root.right);
    }

    private boolean mirror(TreeNode left, TreeNode right) {

        if (left == null && right == null) return true;
        if (left == null || right == null) return false;
        
        return left.val == right.val && mirror(left.left, right.right) && mirror(left.right, right.left);
    }
}

File: src/main/java/leet_code/top_interview_questions_easy/trees/SymmetricTree.go
package main

func main() {

}

type treeNode3 struct {
	Val   int
	Left  *treeNode3
	Right *treeNode3
}

func isSymmetric(root *treeNode3) bool {

	return isMirror(root.Left, root.Right)
}

func isMirror(root1, root2 *treeNode3) bool {
	if root1 == nil && root2 == nil {
		return true
	}
	if root1 == nil || root2 == nil {
		return false
	}
	if root1.Val == root2.Val && isMirror(root1.Left, root2.Right) && isMirror(root1.Right, root2.Left) {
		return true
	}
	return false
}
